core/dbt/cli/flags.py

```diff
@@ -88,3 +95,18 @@ def assign_params(ctx, params_assigned_from_default):
 
     def __str__(self) -> str:
         return str(pf(self.__dict__))
+
+    def _assert_mutually_exclusive(
+        self, params_assigned_from_default: Set[str], group: List[str]
+    ) -> None:
+        """
+        Ensure no elements from group are simultaneously provided by a user, as inferred from params_assigned_from_default.
+        Raises click.UsageError if any two elements from group are simultaneously provided by a user.
+        """
+        set_flag = None
+        for flag in group:
+            flag_set_by_user = flag.lower() not in params_assigned_from_default
+            if flag_set_by_user and set_flag:
+                raise UsageError(f"{flag.lower()}: not allowed with argument {set_flag.lower()}")
```

Comment:
@iknox-fa - click enabled a nice UX improvement over what we've got in main. In main, there wasn't a straight-foward way to avoid dumping a stack trace when we detected mutually exclusive options passed together, but from different sources (e.g. warn_error from cli params, warn_error_options from user config or env). Raising click's UsageError here results in a consistent CLI experience for all of these combos!

For example, 

(warn_error_options set by user config, warn_error by CLI params)

```yaml
# profiles.yml
config:
  warn_error_options:
    include:
      - PackageInstallPathDeprecation
```
```sh 
❯ python3 core/dbt/cli/main.py --warn-error clean --project-dir ~/src/jaffle_shop
Usage: main.py clean [OPTIONS]
Try 'main.py clean -h' for help.

Error: warn_error_options: not allowed with argument warn_error
```

and

(both options set from CLI params)

```sh
❯ python3 core/dbt/cli/main.py --warn-error --warn-error-options '{"include": "all"}' clean --project-dir ~/src/jaffle_shop
Usage: main.py clean [OPTIONS]
Try 'main.py clean -h' for help.

Error: warn_error_options: not allowed with argument warn_error
```

^same error messaging!

cc: @jtcohen6, re: prior discussion in https://github.com/dbt-labs/dbt-core/pull/6520#discussion_r1066303862